### ISSUE TITLE
Synchronous API for position solving

### ISSUE THREAD

**Issue by folz** [floating-ui/floating-ui#2418]
Hi all, thanks for the library! I've been using you for a while, first as popper and I'm excited to see the new API surface you're offering with floating-ui.

## What is the feature you're requesting?
I'd like to have an API that allows for synchronous position solving. That is, instead of the current API of

```ts
computePosition(...).then(position =>)
```

I would be able to use the position value at the moment of the call:

```ts
const position = computePosition(...)
```

## Is your feature request related to a problem? Please describe.
I'd like to have more control over my own ability to track positions of elements outside the scope of floating-ui, while still relying on its position solving implementations. I don't think the solving part needs to be async if I'm able to provide the correct `getBoundingClientRect` virtual elements and platform config synchronously.

For example, `platform.isRTL` could be a `config.isRTL` value that I ensure is up-to-date when I make the `computePosition` call. Then `computePosition` could reference `config.isRTL` as a static value versus dynamically getting it with `await platform.isRTL()`.

Based on some explorations at https://github.com/folz/floating-ui/pull/1, it also looks like the Web implementation doesn't need to be async at all - that seems to be required only because of the react-native platform. If we were able to have a synchronous API for solving, that could unlock a synchronous platform for web, and an asynchronous platform for react-native only.

**Additional context**
See my prototype PR, https://github.com/folz/floating-ui/pull/1, for where I was trying my hand at this. Also, if I'm missing something that's quietly broken about my changes, I'd love to hear more.

I'm also down to help develop this idea and contribute code, but probably no more than a few hours per week til it's done.

**Comment by atomiks** [comment]
It doesn't need to be async on the web but it uses the shared `core` platform which has to be async, so there would be two separate packages (sync and async) which isn't good to maintain. 

Why do you need it to be sync? The positioning is functionally the same except it just gets placed in a microtask.

**Comment by folz** [comment]
I'm building an IDE for livecoding (https://github.com/folz/hiasynth), so there's a lot of UI that I'll have either contextually floating next to some code. I want to build towards a node editor view like TouchDesigner, and for that I'd be managing the positions of those node elements in "node space" but I still want to play nicely with floating tooltips and browser window borders. This means where floating-ui might need async to measure a position, I already know it because my app state positions are the source of truth, and I set the DOM element's dimensions/positions from that.

I'm also exploring ideas around using this sort of positioning for code relationships between expressions - for example, something like "parenting" one expression's node representation to another expression, and keeping those placed next to each other when the user drags the parent element. And again keeping in mind things like "if the parent bumps up against a scene boundary, relocate the child so it's still in the scene"

I think there might be an opportunity to have a `solver-core`, which is sync and takes in the results of the async `platform-core`, which `await`s things like `platform.isRTL` to keep API consistency across all platforms.

It's helpful for me to have this calculated value synchronously because then I don't need to think about how I'll best await this value for the other state stores I'm using, I just set the store directly.

**Comment by folz** [comment]
Also, feel free to lmk this use case is outside the scope of the library and I'm happy to handle it myself. That was going to be my original plan before I noticed the possibility for part of the API to become sync, so I wanted to share the idea with you all.

**Comment by atomiks** [comment]
I don't think it's outside the scope; if you have a suggestion/example of how to keep it maintainable (single source to update) while splitting off async and sync versions, then that's welcome for sure 👍

**Comment by folz** [comment]
Sounds good, I'll give it some thought.

**Comment by tuananhlai** [comment]
I agree that having a synchrous API for `computePosition` would be delightful. I've been using `computePosition` to create a custom hook in React and its async property make it quite hard to test with.
